testing/async/mixin.h

```diff
@@ -0,0 +1,645 @@
+//            typename output_value_type   = input_value_type,
+//            typename initial_value_type  = input_value_type,
+//            typename alternate_binary_op = thrust::maximum<>>
+//   struct basic_invoker
+//      : testing::async::mixin::input::device_vector<input_value_type>
+//      , testing::async::mixin::output::device_vector<output_value_type>
+//      , testing::async::exclusive_scan::mixin::postfix_args::
+//          all_overloads<initial_value_type, alternate_binary_op>
+//      , testing::async::exclusive_scan::mixin::invoke_reference::
+//          host_synchronous<input_value_type, output_value_type>
+//      , testing::async::exclusive_scan::mixin::invoke_async::basic
+//      , testing::async::mixin::compare_outputs::assert_equal_quiet
+//   {
+//     static std::string description()
+//     {
+//       return "basic invocation with device vectors";
+//     }
+//   };
+//
+//   ...
+//
+//   testing::async::test_policy_overloads<basic_invoker<T>>::run(num_values);
+// ```
+//
+// The basic_invoker class expands to something similar to the following:
+//
+// ```
+//  template <typename input_value_type,
+//            typename output_value_type   = input_value_type,
+//            typename initial_value_type  = input_value_type,
+//            typename alternate_binary_op = thrust::maximum<>>
+//  struct basic_invoker
+//  {
+//  public:
+//
+//    static std::string description()
+//    {
+//      return "basic invocation with device vectors";
+//    }
+//
+//    //-------------------------------------------------------------------------
+//    // testing::async::mixin::input::device_vector
+//    using input_type  = thrust::device_vector<input_value_type>;
+//
+//    // Generate an instance of the input:
+//    static input_type generate_input(std::size_t num_values)
+//    {
+//      input_type input(num_values);
+//      thrust::sequence(input.begin(), input.end(), 25, 3);
+//      return input;
+//    }
+//
+//    //-------------------------------------------------------------------------
+//    // testing::async::mixin::output::device_vector
+//    using output_type = thrust::device_vector<output_value_type>;
+//
+//    // Generate an instance of the output:
+//    // Might be more complicated, eg. fancy iterators, etc
+//    static output_type generate_output(std::size_t num_values)
+//    {
+//      return output_type(num_values);
+//    }
+//
+//    //-------------------------------------------------------------------------
+//    //
+//    testing::async::mixin::exclusive_scan::mixin::postfix_args::all_overloads
+//    //
+//    using postfix_args_type = std::tuple<                 // List any extra
+//    arg overloads:
+//      std::tuple<>,                                       // - no extra args
+//      std::tuple<initial_value_type>,                     // - initial_value
+//      std::tuple<initial_value_type, alternate_binary_op> // - initial_value,
+//      binary_op
```

Comment:
`binary_op` appears to have been line wrapped, it should be on the previous line.